### PR TITLE
Allow to set null secret value for Applications if it's public

### LINKED ISSUE #1724: Doorkeeper generates a secret for public clients
Currently doorkeeper's Application model always generates a `secret` value, even when the client type should not have a client secret generated (i.e., public clients). This means we're storing a secret for applications even when a secret value is completely unneeded

I think the model & migrations likely need to be changed to make `oauth_application.secret` nullable, and only generate a secret if `app.confidential?`

The [`Doorkeeper::Application.by_uid_and_secret()`](https://github.com/doorkeeper-gem/doorkeeper/blob/main/lib/doorkeeper/models/application_mixin.rb#L26) method does do the right thing and not bother with secret validation if the client is public.

However, I'm not sure if `POST /oauth/token` will explicitly reject the request if `client_secret` is presented despite the client being public. It seems this happens more through a side effect of `by_uid_and_secret()`'s logic.

This would save significant database resources for applications such as Mastodon, where we've a significant number of applications, due to using a form of Dynamic Client Registration. (i.e., the application count is in the tens or hundreds of thousands, if not millions).

### PR DESCRIPTION
**Pull request by nbulaj**
Aims to fix #1724

### CODE REVIEW

**Review by ThisIsMissEm — APPROVED** [review]
This looks good to me, also, fwiw, I would've been happy to work on a PR, but was a bit stumped by the test failures. With the gemfile change I'll see if I can get the tests running locally with this branch.

**Comment by nbulaj** [comment]
The only thing which makes me worry is legacy apps which already had NOT NULL constraint. If they will add a public app DB will throw an error for NULL secret. I have to think how to make these changes backward compatible.

**Comment by ThisIsMissEm** [comment]
Perhaps we could modify the migration / generator for enabling public clients to check that secret is nullable?

**Comment by nbulaj** [comment]
> Perhaps we could modify the migration / generator for enabling public clients to check that secret is nullable?

You mean add a custom constraint? I already added a check on a model level, but custom constraints are DB-dependant so I;m not sure if we wanna to support all possible variants.

**Comment by ThisIsMissEm** [comment]
I mean, create a migration that can be applied (like adding PKCE) that drops the not-null constrain on client_secret.

File: Gemfile [hunk 1]
```diff
@@ -23,7 +23,7 @@ gem "rubocop-rspec", require: false
 gem "bcrypt", "~> 3.1", require: false
 
 gem "activerecord-jdbcsqlite3-adapter", platform: :jruby
-gem "sqlite3", "~> 2.0", platform: %i[ruby mswin mingw x64_mingw]
+gem "sqlite3", "~> 1.4", platform: [:ruby, :mswin, :mingw, :x64_mingw]
```
[the remark this answers is not in the document]
Comment (nbulaj) [reply Gemfile:26]: Works for me and GitHub CI :thinking: 
Sorry I don't use dev-containers / docker / other local setups, not sure what can happen there

[the remark this answers is not in the document]
Comment (ThisIsMissEm) [reply Gemfile:26]: I don't either!

**Pull request merged**